application_manager/plugins/spark_mesos/plugin.py

```diff
@@ -0,0 +1,250 @@
+                            % (time.strftime("%H:%M:%S")))
+
+        i, o, e = conn.exec_command(spark_run % (api.spark_path,
+                                                 self.app_id,
+                                                 api.mesos_url,
+                                                 api.mesos_port,
+                                                 execution_class,
+                                                 binary_path,
+                                                 execution_parameters))
+
+        # Discovery ips of the executors from Mesos
+        # and discovery the ids on KVM using the ips
+        list_vms_one = 'onevm list --user %s --password %s --endpoint %s' % \
+                       (api.one_username, api.one_password, api.one_url)
+        stdin, stdout, stderr = conn.exec_command(list_vms_one)
+
+        list_response = stdout.read()
+        vms_ips, master = self._get_executors_ip(conn)
+        vms_ids = self._extract_vms_ids(list_response)
+
+        executors_vms_ids = []
+        for ip in vms_ips:
+            for id in vms_ids:
+                vm_info_one = 'onevm show %s --user %s --password %s \
+                              --endpoint %s' % (id, api.one_username,
+                                                api.one_password, api.one_url)
+
+                stdin, stdout, stderr = conn.exec_command(vm_info_one)
+                if ip in stdout.read():
+                    executors_vms_ids.append(id)
+                    break
+
+        plugin_log.log("%s | Executors IDs: %s"
+                        % (time.strftime("%H:%M:%S"), executors_vms_ids))
+
+        # Set up the initial configuration of cpu cap
+        scaler.setup_environment(api.controller_url, executors_vms_ids,
+                                 starting_cap, data)
+
+        info_plugin = {"spark_submisson_url": master,
+                       "expected_time": expected_time,
+                       "number_of_jobs": number_of_jobs}
+
+        plugin_log.log("%s | Starting monitor" % (time.strftime("%H:%M:%S")))
+        monitor.start_monitor(api.monitor_url, self.app_id,
+                              'spark-mesos', info_plugin, 2)
+
+        plugin_log.log("%s | Starting scaler" % (time.strftime("%H:%M:%S")))
+        scaler.start_scaler(api.controller_url,
+                            self.app_id,
+                            executors_vms_ids,
+                            data)
+
+        # This command locks the plugin execution until the execution be done
+        print o.read()
+
+        plugin_log.log("%s | Stopping monitor" % (time.strftime("%H:%M:%S")))
+        monitor.stop_monitor(api.monitor_url, self.app_id)
+
+        plugin_log.log("%s | Stopping scaler" % (time.strftime("%H:%M:%S")))
+        scaler.stop_scaler(api.controller_url, self.app_id)
+
+        plugin_log.log("%s | Remove binaries" % (time.strftime("%H:%M:%S")))
+        conn.exec_command('rm -rf ~/exec_bin.*')
+        return None, None
```

Comment:
Why are you returning a tuple of None objects? What is the caller expecting as result?

Reply:
The return of 'execute' method of an plugin provider should be the application id and the object of executor.